VIX-2940 Added the ability to rotate all the props in Preview.

da-markson opened the pull request.

johncbaur commented on src/Vixen.Modules/Preview/VixenPreview/Shapes/PreviewLine.cs:
Maybe this is by design but if I rotate a line and then move one of the points.  The rotation point is offset the line.

johncbaur commented on src/Vixen.Modules/Preview/VixenPreview/Shapes/PreviewBaseShape.cs:
```diff
@@ -57,6 +63,17 @@ public string Name
 		[Browsable(false)]
 		public virtual string TypeName => @"Shape";
 
+		public virtual void OnDeserialized(StreamingContext context)
```
This method doesn't seem to fit the normal virtual / override scenarios.   I am reading that the best practice might be to use override in the derived classes and call the base class.

da-markson replied:
You lost me a little bit on what your asking me to change? If I return the base class to abstract then I can't add any body.  Can you please elaborate a little?

johncbaur replied:
What I noticed here is that the base class is marked as virtual so I would expect all derived shapes to include the keyword override in their version of the method.

johncbaur replied:
I think it should look like this:
PreviewBaseShape.cs - public virtual void OnDeserialized(StreamingContext context)
PreviewLightBaseShape.cs - public override void OnDeserialized(StreamingContext context)
PreviewCane.cs - public override void OnDeserialized(StreamingContext context)

You can still call the base before calling Layout.  Does the above avoid the compile error you were seeing?

johncbaur commented on src/Vixen.Modules/Preview/VixenPreview/Shapes/PreviewBaseShape.cs:
```diff
@@ -3,16 +3,22 @@
-		
+		public int _rotationAngle = 0;
+		protected PreviewPoint _rotateHandle;
+		public PreviewPoint _rotationAxis;
```
I don't think the team has strict coding standards but _ underscore should denote a private field.  Consider changing RotationAxis and RotationAngle to be a property with the Browsable false tag.  _rotateHandle should also be a protected property without the underscore in my opinion.

da-markson replied:
1) I removed the leading underscore from _rotationAxis, _rotationAngle and _rotationHandle 
2) I added Browseable(false) tag for rotationAxis and rotationAngle

johncbaur replied:
These files already had issues with respect to industry coding standards, so I am not demanding that you fix the issues for these new class members them but at the same time developing good practices moving forward keeps the code base clean and consistent.  So your updates were moving in the correct direction but not quite correct/complete in my opinion.

The rule I operate by is that the only allowable field is a private field with the underscore.  Anything that is public or protected should be a property and not a field.  This allows for encapsulation if there are additional behavior required when changing the member variable.
 
[Browsable(false)]
public int RotationAngle { get; set; } = 0;

[Browsable(false)]
public PreviewPoint RotationAxis { get; set; }

jeffu231 replied:
> I don't think the team has strict coding standards but _ underscore should denote a private field. Consider changing RotationAxis and RotationAngle to be a property with the Browsable false tag. _rotateHandle should also be a protected property without the underscore in my opinion.

We don't have strict standards, but we should try to follow generally accepted coding practices for new code. I agree with the suggestions John has made.  I know the current code has some variances in that, especially from the early days. But if we are working on a section of code and there is an opportunity to standardize things, I am not opposed. Similar to fixing spaces vs tabs or other formatting, if the change is strictly formatting, I would prefer you to isolate it from logic changes.

da-markson commented on src/Vixen.Common/Resources/Resources.csproj:
```diff
@@ -20,6 +20,7 @@
 				<None Include="attributes_display.png" />
 				<None Include="bullet_black.png" />
 				<None Include="bullet_blue.png" />
+				<None Include="C:\Users\david\Downloads\DoubleRotate.ico" Link="DoubleRotate.ico" />
```
Corrected.

da-markson commented on src/Vixen.Modules/Preview/VixenPreview/Shapes/PreviewIcicle.cs:
If you rotate the icicles 90 degrees, than the icicles run horizontally.  Why is this wrong? Granted, it would be the classical meaning of icicles (gravity drips down), but why wouldn't this be potentially useful in some other way?

Of course if you want, I can absolutely disable this.

johncbaur replied:
To argue against myself, I guess if the user was using something like Boscoyo icicle coro the rotation would allow them to orient the coro going up an eave.  So maybe there is a use case to including it.

jeffu231 replied:
> To argue against myself, I guess if the user was using something like Boscoyo icicle coro the rotation would allow them to orient the coro going up an eave. So maybe there is a use case to including it.

I don't think it hurts giving the use flexibility to rotate any of them. I think the key think to emphasis in documentation and anywhere we can is rotating is not a fix for improperly linked/patched props. As we have seen in the past they try to rotate things that appear upside down when in actuality the patching is incorrect.

da-markson commented on src/Vixen.Modules/Preview/VixenPreview/Shapes/PreviewCustomProp.cs:
I tested this with the few custom props I have and this doesn't fail. Can you send me this prop so I can get to the bottom of this error

johncbaur replied:
I attached the prop to the JIRA ticket.  If you can't produce the problem I can attach the debugger and see what I can observe.  It seemed to crash on mouse release.

johncbaur replied:
What I am seeing is that in TransformPreviewPoint in PreviewTool.cs, line 246 the shape._selectedPoints is null.  That is the selectedPoints is null.  I think the key to reproducing this issue to start the rotation and then move the mouse towards the center of the prop.  The cursor changes to a move a cursor.  At which point when I release the mouse button it crashes.

da-markson commented on src/Vixen.Modules/Preview/VixenPreview/Shapes/PreviewTools.cs:
```diff
@@ -165,6 +169,183 @@ public static List<Point> GetEllipsePoints(
+		{
+			// Check to see if the shape is inverted (i.e. the Topleft is above the BottomRight).  If so, then set the rotational
+			// origin to 180 degrees since all the calulations need to be rotated by 180 degrees.
+			var _topleft = shape._selectPoints.Find(x => x.PointType == PreviewPoint.PointTypes.SizeTopLeft);
```
Removed the underscore from _topLeft

da-markson commented:
Hmmm.  I can rotate, size, and translate it with no problem.  Yes, if you
could provide me with more info, I would greatly appreciate it.

Thanks,

David Markson
*Cell*:        (817) 584-4755
*E-mail*:    ***@***.***




On Sun, Feb 23, 2025 at 9:07 PM johncbaur ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> On src/Vixen.Modules/Preview/VixenPreview/Shapes/PreviewCustomProp.cs
> <https://github.com/VixenLights/Vixen/pull/681#discussion_r1966999220>:
>
> I attached the prop to the JIRA ticket. If you can't produce the problem I
> can attach the debugger and see what I can observe. It seemed to crash on
> mouse release.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/VixenLights/Vixen/pull/681#discussion_r1966999220>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BMJ3GZUQNTFGMFC5UHSDVA32RKEHNAVCNFSM6AAAAABXWZOKWGVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDMMZVHE2DKMJQHE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

da-markson commented:
I have absolutely no problem with making these changes.  Because
rotationAngle is the field value for RotationAngle on line 131, I'll need
to change rotationAngle to rotateAngle.

Thanks,

David Markson
*Cell*:        (817) 584-4755
*E-mail*:    ***@***.***




On Sun, Feb 23, 2025 at 9:31 PM johncbaur ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/Vixen.Modules/Preview/VixenPreview/Shapes/PreviewBaseShape.cs
> <https://github.com/VixenLights/Vixen/pull/681#discussion_r1967007544>:
>
> >  	public abstract class PreviewBaseShape : ICloneable, IDisposable
>  	{
>  		public string _name;
>  		protected  bool _selected = false;
> -		
> +		public int _rotationAngle = 0;
> +		protected PreviewPoint _rotateHandle;
> +		public PreviewPoint _rotationAxis;
>
> These files already had issues with respect to industry coding standards,
> so I am not demanding that you fix the issues for these new class members
> them but at the same time developing good practices moving forward keeps
> the code base clean and consistent. So your updates were moving in the
> correct direction but not quite correct/complete in my opinion.
>
> The rule I operate by is that the only allowable field is a private field
> with the underscore. Anything that is public or protected should be a
> property and not a field. This allows for encapsulation if there are
> additional behavior required when changing the member variable.
>
> [Browsable(false)]
> public int RotationAngle { get; set; } = 0;
>
> [Browsable(false)]
> public PreviewPoint RotationAxis { get; set; }
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/VixenLights/Vixen/pull/681#discussion_r1967007544>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BMJ3GZXF6UOGQEMGB63N4GL2RKG7XAVCNFSM6AAAAABXWZOKWGVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDMMZVHE2TSMBWGQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

da-markson commented:
The base class is marked as abstract.

If I change the specific method from
     private new void OnDeserialized(StreamingContext context)
to
     public override void OnDeserialized(StreamingContext context)

the compiler complains with a CS0506 error.

Additionally, I need to call the base class before the derived class
completes (before the layout() call), so I need a little more control.


Thanks,

David Markson
*Cell*:        (817) 584-4755
*E-mail*:    ***@***.***




On Sun, Feb 23, 2025 at 8:57 PM johncbaur ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/Vixen.Modules/Preview/VixenPreview/Shapes/PreviewBaseShape.cs
> <https://github.com/VixenLights/Vixen/pull/681#discussion_r1966996189>:
>
> > @@ -57,6 +63,17 @@ public string Name
>  		[Browsable(false)]
>  		public virtual string TypeName => @"Shape";
>
> +		public virtual void OnDeserialized(StreamingContext context)
>
> What I noticed here is that the base class is marked as virtual so I would
> expect all derived shapes to include the keyword override in their version
> of the method.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/VixenLights/Vixen/pull/681#discussion_r1966996189>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BMJ3GZUJJCBFSECOLR4WNVL2RKDDHAVCNFSM6AAAAABXWZOKWGVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDMMZVHEZTSOJYGI>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

da-markson commented:
What black magic are you doing?  I can not reproduce this AT ALL!

Regardless, I added a check for null near that line so it should avoid your
issue, but I sure would prefer to know how to trip into this fault.

Thanks,

David Markson
*Cell*:        (817) 584-4755
*E-mail*:    ***@***.***




On Sun, Feb 23, 2025 at 9:57 PM johncbaur ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> On src/Vixen.Modules/Preview/VixenPreview/Shapes/PreviewCustomProp.cs
> <https://github.com/VixenLights/Vixen/pull/681#discussion_r1967017137>:
>
> What I am seeing is that in TransformPreviewPoint in PreviewTool.cs, line
> 246 the shape._selectedPoints is null. That is the selectedPoints is null.
> I think the key to reproducing this issue to start the rotation and then
> move the mouse towards the center of the prop. The cursor changes to a move
> a cursor. At which point when I release the mouse button it crashes.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/VixenLights/Vixen/pull/681#discussion_r1967017137>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BMJ3GZXIJ5I4L4LIJURSDWL2RKKDXAVCNFSM6AAAAABXWZOKWGVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDMMZVHE3TGNBXGY>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

da-markson commented:
Additionally, thinking about this further, how can you have no
selectpoints since you are rotating?  You should have at least four?

Thanks,

David Markson
*Cell*:        (817) 584-4755
*E-mail*:    ***@***.***




On Sun, Feb 23, 2025 at 10:41 PM David Markson ***@***.***>
wrote:

> What black magic are you doing?  I can not reproduce this AT ALL!
>
> Regardless, I added a check for null near that line so it should avoid
> your issue, but I sure would prefer to know how to trip into this fault.
>
> Thanks,
>
> David Markson
> *Cell*:        (817) 584-4755
> *E-mail*:    ***@***.***
>
>
>
>
> On Sun, Feb 23, 2025 at 9:57 PM johncbaur ***@***.***>
> wrote:
>
>> ***@***.**** commented on this pull request.
>> ------------------------------
>>
>> On src/Vixen.Modules/Preview/VixenPreview/Shapes/PreviewCustomProp.cs
>> <https://github.com/VixenLights/Vixen/pull/681#discussion_r1967017137>:
>>
>> What I am seeing is that in TransformPreviewPoint in PreviewTool.cs, line
>> 246 the shape._selectedPoints is null. That is the selectedPoints is null.
>> I think the key to reproducing this issue to start the rotation and then
>> move the mouse towards the center of the prop. The cursor changes to a move
>> a cursor. At which point when I release the mouse button it crashes.
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/VixenLights/Vixen/pull/681#discussion_r1967017137>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/BMJ3GZXIJ5I4L4LIJURSDWL2RKKDXAVCNFSM6AAAAABXWZOKWGVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDMMZVHE3TGNBXGY>
>> .
>> You are receiving this because you authored the thread.Message ID:
>> ***@***.***>
>>
>

da-markson commented:
No, I do not understand why.  I am trying your concept of first creating
the prop in a new profile in the release version, but am still not able to
reproduce.  I traced through the code and I can not yet find the path where
_selectPoints is null.

I also did a Clean and Rebuild in case that was the issue, but no luck. I
am at a lost as to why I can't reproduce such an obvious problem,


Thanks,

David Markson
*Cell*:        (817) 584-4755
*E-mail*:    ***@***.***




On Mon, Feb 24, 2025 at 8:07 PM johncbaur ***@***.***> wrote:

> ***@***.**** requested changes on this pull request.
> ------------------------------
>
> In src/Vixen.Modules/Preview/VixenPreview/Shapes/PreviewTools.cs
> <https://github.com/VixenLights/Vixen/pull/681#discussion_r1968703238>:
>
> > @@ -235,7 +235,7 @@ public static PreviewPoint TransformPreviewPoint(PreviewBaseShape shape, Preview
>  				shape._selectedPoint.PointType = PreviewPoint.PointTypes.RotateHandle;
>
>  				// Check to see if any translation is needed.
> -				if (newCenter.X != shape.Center.X || newCenter.Y != shape.Center.Y)
> +				if (shape._selectPoints != null && (newCenter.X != shape.Center.X || newCenter.Y != shape.Center.Y))
>
> Do we understand why _selectPoints is null for an existing profile? I have
> not spent the time to understand all of this logic but hopefully the
> additional information I provided on how to reproduce the crash might give
> you more insight.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/VixenLights/Vixen/pull/681#pullrequestreview-2638929736>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BMJ3GZU6MIDBEJTZO7NDCKL2RPF5LAVCNFSM6AAAAABXWZOKWGVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDMMZYHEZDSNZTGY>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

da-markson commented:
I FINALLY got it to fail.  Working on a fix now.

Thanks,

David Markson
*Cell*:        (817) 584-4755
*E-mail*:    ***@***.***




On Mon, Feb 24, 2025 at 8:19 PM David Markson ***@***.***>
wrote:

> No, I do not understand why.  I am trying your concept of first creating
> the prop in a new profile in the release version, but am still not able to
> reproduce.  I traced through the code and I can not yet find the path where
> _selectPoints is null.
>
> I also did a Clean and Rebuild in case that was the issue, but no luck. I
> am at a lost as to why I can't reproduce such an obvious problem,
>
>
> Thanks,
>
> David Markson
> *Cell*:        (817) 584-4755
> *E-mail*:    ***@***.***
>
>
>
>
> On Mon, Feb 24, 2025 at 8:07 PM johncbaur ***@***.***>
> wrote:
>
>> ***@***.**** requested changes on this pull request.
>> ------------------------------
>>
>> In src/Vixen.Modules/Preview/VixenPreview/Shapes/PreviewTools.cs
>> <https://github.com/VixenLights/Vixen/pull/681#discussion_r1968703238>:
>>
>> > @@ -235,7 +235,7 @@ public static PreviewPoint TransformPreviewPoint(PreviewBaseShape shape, Preview
>>  				shape._selectedPoint.PointType = PreviewPoint.PointTypes.RotateHandle;
>>
>>  				// Check to see if any translation is needed.
>> -				if (newCenter.X != shape.Center.X || newCenter.Y != shape.Center.Y)
>> +				if (shape._selectPoints != null && (newCenter.X != shape.Center.X || newCenter.Y != shape.Center.Y))
>>
>> Do we understand why _selectPoints is null for an existing profile? I
>> have not spent the time to understand all of this logic but hopefully the
>> additional information I provided on how to reproduce the crash might give
>> you more insight.
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/VixenLights/Vixen/pull/681#pullrequestreview-2638929736>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/BMJ3GZU6MIDBEJTZO7NDCKL2RPF5LAVCNFSM6AAAAABXWZOKWGVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDMMZYHEZDSNZTGY>
>> .
>> You are receiving this because you authored the thread.Message ID:
>> ***@***.***>
>>
>

da-markson commented:
You were correct that the Prop MUST be created in the earlier version to
trigger this fault.  That was the missing ingredient where I couldn't
replicate.

I've addressed this failure and pushed the code.  When you have the
opportunity, please evaluate to validate that I've addressed the issue.

Thanks,

David Markson
*Cell*:        (817) 584-4755
*E-mail*:    ***@***.***




On Mon, Feb 24, 2025 at 8:07 PM johncbaur ***@***.***> wrote:

> ***@***.**** requested changes on this pull request.
> ------------------------------
>
> In src/Vixen.Modules/Preview/VixenPreview/Shapes/PreviewTools.cs
> <https://github.com/VixenLights/Vixen/pull/681#discussion_r1968703238>:
>
> > @@ -235,7 +235,7 @@ public static PreviewPoint TransformPreviewPoint(PreviewBaseShape shape, Preview
>  				shape._selectedPoint.PointType = PreviewPoint.PointTypes.RotateHandle;
>
>  				// Check to see if any translation is needed.
> -				if (newCenter.X != shape.Center.X || newCenter.Y != shape.Center.Y)
> +				if (shape._selectPoints != null && (newCenter.X != shape.Center.X || newCenter.Y != shape.Center.Y))
>
> Do we understand why _selectPoints is null for an existing profile? I have
> not spent the time to understand all of this logic but hopefully the
> additional information I provided on how to reproduce the crash might give
> you more insight.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/VixenLights/Vixen/pull/681#pullrequestreview-2638929736>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BMJ3GZU6MIDBEJTZO7NDCKL2RPF5LAVCNFSM6AAAAABXWZOKWGVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDMMZYHEZDSNZTGY>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

jeffu231 reviewed (commented):
I agree this solution will allow both methods to be called because the serialization process calls both methods due to the annotation. You can access a hidden member of the base class by calling it with the base keyword, which is likely what the serializer is doing. I would think this is a good place for some comments to add clarity to what is occurring so in the future we can remember this discussion and understand what is occurring.

jeffu231 reviewed (approved):
Feels like we have gone over this one pretty good. 

@johncbaur  Any further remarks from you? If not I think we can merge this one up.
